doc/website/blog: ogent announcement (#2327)
ent/ent: doc/website/blog/2022-02-15-generate-rest-crud-with-ent-and-ogen.md

---
title: Auto generate REST CRUD with Ent and ogen 
author: MasseElch 
authorURL: "https://github.com/masseelch"
authorImageURL: "https://avatars.githubusercontent.com/u/12862103?v=4"
image: "https://entgo.io/images/assets/elkopa/entoas-code.png"
---

> End of 2021 we announced, that [Ent](https://entgo.io) got a new official extension to generate a fully

Reviewer: ```suggestion
In the end of 2021 we announced that [Ent](https://entgo.io) got a new official extension to generate a fully
```

compliant [OpenAPI Specification](https://swagger.io/resources/open-api/)
document: [`entoas`](https://github.com/ent/contrib/tree/master/entoas).

> Today, we are very happy to announce, that there is a new extension built to work

Reviewer: ```suggestion
Today, we are very happy to announce that there is a new extension built to work
```

with `entoas`: [`ogent`](https://github.com/ariga/ogent). It utilizes the power
of [`ogen`](https://github.com/ogen-go/ogen) ([website](https://ogen.dev/docs/intro/)) to provide a type-safe,

> reflection free implementation of the OpenAPI Specification document generated by `entoas`. 

Reviewer: ```suggestion
reflection-free implementation of the OpenAPI Specification document generated by `entoas`. 
```

```shell
go get entgo.io/contrib/entoas@master ariga.io/ogent@master
```

Now follow the next two steps to enable them and to configure Ent to work with the extensions:

1\. Create a new Go file named `ent/entc.go` and paste the following content:

```go title="ent/entc.go"
//go:build ignore

package main

import (
	"log"

	"entgo.io/contrib/entoas"
	"entgo.io/ent/entc"
	"entgo.io/ent/entc/gen"
	"github.com/ariga/ogent"
	"github.com/ogen-go/ogen"
)

func main() {
	spec := new(ogen.Spec)
	oas, err := entoas.NewExtension(entoas.Spec(spec))
	if err != nil {
		log.Fatalf("creating entoas extension: %v", err)
	}
	ogent, err := ogent.NewExtension(spec)
	if err != nil {
		log.Fatalf("creating ogent extension: %v", err)
	}
	err = entc.Generate("./schema", &gen.Config{}, entc.Extensions(ogent, oas))
	if err != nil {
		log.Fatalf("running ent codegen: %v", err)
	}
}
```

2\. Edit the `ent/generate.go` file to execute the `ent/entc.go` file:

```go title="ent/generate.go"
package ent

//go:generate go run -mod=mod entc.go
```

With these steps complete, all is set up for generating an OAS document and implementing server code from your schema!

### Generate a CRUD HTTP API Server

> The first step on our way to the HTTP API server is to create an Ent schema graph. For the sake of brevity here is an

Reviewer: ```suggestion
The first step on our way to the HTTP API server is to create an Ent schema graph. For the sake of brevity, here is an
```

example schema to use:

```go title="ent/schema/todo.go"
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/schema/field"
)

// Todo holds the schema definition for the Todo entity.
type Todo struct {
	ent.Schema
}

// Fields of the Todo.
func (Todo) Fields() []ent.Field {
	return []ent.Field{
		field.String("title"),
		field.Bool("done"),
	}
}
```

> The code above is the Ent-way to describe a schema-graph. In this particular case we created a todo entity.

Reviewer: ```suggestion
The code above is the "Ent way" to describe a schema-graph. In this particular case we created a todo entity.
```

				Errors: rawError(err),
			}, nil
		default:
			// Let the server handle the error.
			return nil, err
		}
	}
	return NewTodoRead(e), nil
}
```

### Run the server

The next step is to create a `main.go` file and wire up all the ends to create an application-server to serve the
Todo-API. The following main function initializes a SQLite in-memory database, runs the migrations to create all the
tables needed and serves the API as described in the `ent/openapi.json` file on `localhost:8080`: 

```go title="main.go"
package main

import (
	"context"
	"log"
	"net/http"

	"entgo.io/ent/dialect"
	"entgo.io/ent/dialect/sql/schema"
	"<your-project>/ent"
	"<your-project>/ent/ogent"
	_ "github.com/mattn/go-sqlite3"
)

func main() {
	// Create ent client.
	client, err := ent.Open(dialect.SQLite, "file:ent?mode=memory&cache=shared&_fk=1")
	if err != nil {
		log.Fatal(err)
	}
	// Run the migrations.
	if err := client.Schema.Create(context.Background(), schema.WithAtlas(true)); err != nil {
		log.Fatal(err)
	}
	// Start listening.
	srv := ogent.NewServer(ogent.NewOgentHandler(client))
	if err := http.ListenAndServe(":8080", srv); err != nil {
		log.Fatal(err)
	}
}
```

> After you run the server with `go run -mod=mod main.go` you can work with the API. First let's create a new Todo. For

Reviewer: ```suggestion
After you run the server with `go run -mod=mod main.go` you can work with the API. 

First, let's create a new Todo. For
```

demonstration purpose we do not send a request body:

```shell
↪ curl -X POST -H "Content-Type: application/json" localhost:8080/todos
{
  "error_message": "body required"
}
```

> As you can see `ogen` handles that case for you since `entoas` did mark the body as required when attempting to create a

Reviewer: ```suggestion
As you can see `ogen` handles that case for you since `entoas` marked the body as required when attempting to create a
```

> new resource. Let's try again but this time provide a request body:

Reviewer: ```suggestion
new resource. Let's try again, but this time provide a request body:
```

```shell
↪ curl -X POST -H "Content-Type: application/json" -d '{"title":"Give ogen and ogent a Star on GitHub"}'  localhost:8080/todos
{
  "error_message": "decode CreateTodo:application/json request: invalid: done (field required)"
}
```

> Ooops! What did go wrong? `ogen` has your back: the field `done` is required. To fix this head over to your schema

Reviewer: ```suggestion
Ooops! What went wrong? `ogen` has your back: the field `done` is required. To fix this head over to your schema
```

definition and mark the done field as optional:

```go {18} title="ent/schema/todo.go"
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/schema/field"
)

// Todo holds the schema definition for the Todo entity.
type Todo struct {
	ent.Schema
}

// Fields of the Todo.
func (Todo) Fields() []ent.Field {
	return []ent.Field{
		field.String("title"),
		field.Bool("done").
		    Optional(),
	}
}
```

> Since we made a change to our configuration we have to re-run code generation and restart the server:

Reviewer: ```suggestion
Since we made a change to our configuration, we have to re-run code generation and restart the server:
```

```shell
go generate ./...
go run -mod=mod main.go
```

Now, if we attempt to create the Todo again, see what happens:

```shell
↪ curl -X POST -H "Content-Type: application/json" -d '{"title":"Give ogen and ogent a Star on GitHub"}'  localhost:8080/todos
{
  "id": 1,
  "title": "Give ogen and ogent a Star on GitHub",
  "done": false
}
```

> Voila, there is a new Todo item in the database.

Reviewer: ```suggestion
Voila, there is a new Todo item in the database!
```

Assume you have completed your Todo and starred both [`ogen`](https://github.com/ogen-go/ogen)
and [`ogent`](https://github.com/ariga/ogent) (**you really should!**), mark the todo as done by raising a PATCH
request:

```shell
↪ curl -X PATCH -H "Content-Type: application/json" -d '{"done":true}'  localhost:8080/todos/1
{
  "id": 1,
  "title": "Give ogen and ogent a Star on GitHub",
  "done": true
}
```

### Add custom endpoints

> As you can see the Todo is now marked as done. It would be cooler though to have an extra route for marking a Todo as

Reviewer: ```suggestion
As you can see the Todo is now marked as done. Though it would be cooler to have an extra route for marking a Todo as
```

> done: `PATCH todos/:id/done`. To make this happen we have to do two things: Document the new route in our OAS document

Reviewer: ```suggestion
done: `PATCH todos/:id/done`. To make this happen we have to do two things: document the new route in our OAS document
```
